analysis/blacklist/hostnames.go

```diff
@@ -0,0 +1,124 @@
+package blacklist
+
+import (
+	"unsafe"
+
+	"github.com/ocmdev/rita-bl/list"
+	"github.com/ocmdev/rita/database"
+	"github.com/ocmdev/rita/datatypes/dns"
+	"github.com/ocmdev/rita/datatypes/structure"
+
+	bl "github.com/ocmdev/rita-bl"
+	data "github.com/ocmdev/rita/datatypes/blacklist"
+	log "github.com/sirupsen/logrus"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type hostnameShort struct {
+	Host string `bson:"host"`
+}
+
+//buildBlacklistedHostnames builds a set of blacklisted hostnames from the
+//iterator provided, the system config, a handle to rita-blacklist,
+//and a buffer of hostnames to check at a time
+func buildBlacklistedHostnames(hostnames *mgo.Iter, res *database.Resources,
+	blHandle *bl.Blacklist, bufferSize int) {
+	//create session to write to
+	ssn := res.DB.Session.Copy()
+	defer ssn.Close()
+
+	outputCollection := ssn.DB(res.DB.GetSelectedDB()).C(
+		res.System.BlacklistedConfig.HostnamesTable,
+	)
+	//create type for communicating rita-bl results
+	resultsChannel := make(resultsChan)
+
+	go checkRitaBlacklistHostnames(hostnames, blHandle, bufferSize, resultsChannel)
+
+	//results are maps from ip addresses to arrays of their respective results
+	for results := range resultsChannel {
+		//loop over the map
+		for hostname, individualResults := range results {
+			//if the hostname has blacklist results
+			if len(individualResults) > 0 {
+				blHostname := data.BlacklistedHostname{Hostname: hostname}
+				for _, result := range individualResults {
+					blHostname.Lists = append(blHostname.Lists, result.List)
+				}
+				err := fillBlacklistedHostname(
+					&blHostname,
+					res.DB.GetSelectedDB(),
+					res.System.DNSConfig.HostnamesTable,
+					res.System.StructureConfig.UniqueConnTable,
+					ssn,
+				)
+				if err != nil {
+					res.Log.WithFields(log.Fields{
+						"err":      err.Error(),
+						"hostname": hostname,
+						"db":       res.DB.GetSelectedDB(),
+					}).Error("could not aggregate info on blacklisted hostname")
+					continue
+				}
+				outputCollection.Insert(&blHostname)
+			}
+		}
+	}
+}
+
+func checkRitaBlacklistHostnames(hostnames *mgo.Iter, blHandle *bl.Blacklist,
+	bufferSize int, resultsChannel resultsChan) {
+	i := 0
+	//read in bufferSize entries and check them. Then ship them off to the writer
+	var buff = make([]hostnameShort, bufferSize)
+	for hostnames.Next(&buff[i]) {
+		if i == bufferSize-1 {
+			//see comment in checkRitaBlacklistIPs
+			indexesArray := (*[]string)(unsafe.Pointer(&buff))
```

Comment:
This structure is being treated as an array of strings, but if that ever changes, this will break. Will we ever need to add fields to this structure?